### ISSUE TITLE
Access to "SetPartitionsAssignedHandler" and "SetPartitionsRevokedHandler"

### ISSUE THREAD

**Issue by msallin** [BEagle1984/silverback#34]
I need to get notified if the mentioned events happen.

As far as I see there is currently no possibility to set those handlers.
https://github.com/BEagle1984/silverback/blob/develop/src/Silverback.Integration.Kafka/Messaging/Broker/InnerConsumerWrapper.cs

As this is a blocker for my current project, I will gladly submit a PR. What do you think are good ways to expose access those handlers?

**Comment by BEagle1984** [comment]
This is an interesting requirement. It's gonna be a bit challenging to implement it in a usable way, while still keeping the interface generic (to allow other message brokers to be added in the future).

A possibility could be to expose some events in the KafkaBroker (https://github.com/BEagle1984/silverback/blob/develop/src/Silverback.Integration.Kafka/Messaging/Broker/KafkaBroker.cs).

So to access them you could do something like this:
```
var kafkaBroker = serviceProvider.GetService<IBroker>() as KafkaBroker;

kafkaBroker.PartitionAssigned += ...;
kafkaBroker.PartitionRevoked += ...;
```

Of course the idea is to relay all events from all consumers to through the KafkaBroker's events.

What do you think?

**Comment by msallin** [comment]
This is one possibility but it smells, as if you exchange the IBroker implementation it will break (e.g. with the InMemory equivalent).

A possible approach would be the use of dependency injection. In `Silverback.Integration.Kafka` define two interfaces `IPartitionAssignedHandler` and `IPartitionRevokedHandler`. Those are non-required services. If I want to have code executed, I create a class which implements one of the interfaces and register it on the DI container.

It gets a bit more difficult when its about consuming them. I see three options.
1) Use IServiceProvider in thee KafkaBroker
2) Use the Null-Object Pattern and register Null-Object implementations of those interfaces by default (to be overwritten)
3) Keep the KafkaBroker free from IServiceProvider but implement a factory which will take the reference to the IServiceProvider - this could also be in combination with the option 2.

**Comment by BEagle1984** [comment]
Yep, another option could otherwise be to publish some events to the internal bus like `KafkaPartitionAssignedEvent`, `KafkaPartitionRevokedEvent`, etc.  So that you can subscribe through the usual mechanism of your choice.

**Comment by msallin** [comment]
Hmm.. I like the approach with the events! The only thing here is, that we might mess up the order, don't we? E.g. `onPartitionsRevoked` is invoked right before you loose control of the partition. That means you can e.g. commit. Are the events sync?

**Comment by BEagle1984** [comment]
The messages that we publish via IPublisher are being processed in a synchronous fashion. In other words IPublisher.Publish/PublishAsync returns when all subscribers have returned. So it should work.

I think it should work!

**Comment by BEagle1984** [comment]
Implemented in [1.2.0](https://github.com/BEagle1984/silverback/releases/1.2.0)
